### ISSUE TITLE
Feature request: Flush Plugin Caches with Silence

### ISSUE THREAD

**Issue by GhostNaN** [wwmm/easyeffects#2938]
For a while now I have been experiencing popping/crackling when starting a new audio stream.
This issue is already well known as seen here https://github.com/wwmm/easyeffects/issues/1856 and here https://github.com/wwmm/easyeffects/issues/1392
So I understand that the issue is: 
> plugins you are using still has some of the old data saved internally when you play something new.

After doing my own testing, I can also confirm running this:
```aplay -r 48000 -f S24_3LE /dev/zero```  as suggested by someone seems to "fix" this issue.
Although, it's less than ideal to just have a silent audio stream constantly playing in the background. 

So would it be possible to implement some sort of feature to just play silence after a stream ends?
For example, having a setting in preferences that just adds 50 ms of silence or something?
The idea is just to flush the cache of the "old data" to avoid this issue.

**Comment by Digitalone1** [comment]
Did you already enable the inactivity time option with a huge number (like 1 hour)?

**Comment by GhostNaN** [comment]
```30s```, ```off```, ```3600s`` no change.
It's not because is easyeffects is going inactive.

**Comment by wwmm** [comment]
>30s, off, 3600s no change.
>It's not because easyeffects is going inactive.

Run `pw-dot` so we can see each link state.  Its output file can be viewed with the command `xdot`. If the links are there and still in the active state then EE is already playing silence. But there are situations where PipeWire may have paused the pipeline on its own. We create passive links so it can do that when possible.

>So would it be possible to implement some sort of feature to just play silence after a stream ends?
For example, having a setting in preferences that just adds 50 ms of silence or something?
The idea is just to flush the cache of the "old data" to avoid this issue.

There isn't an elegant or straightforward way to do that. PipeWire is the one managing each filter state and there isn't a pipeline object with flushing mechanisms like in GStreamer. What we could try is some kind of hack. Creating a node like the one from the test signal that plays silence and linking it to our virtual sink. IF that is going to have some unexpected side effect is hard to say without trying it.

**Comment by wwmm** [comment]
>But there are situations where PipeWire may have paused the pipeline on its own. 

It always does it on its own. What I meant is that you could have done that before we usually remove the links between the filters. The removal of the links is what the timeout controls.

**Comment by GhostNaN** [comment]
> It always does it on its own. What I meant is that it could have done that before we usually remove the links between the filters. 

Like when a node suspends with  ```["session.suspend-timeout-seconds"] = 5``` ?

> The removal of the links is what the timeout controls.

I have been using ```qpwgraph``` to visualize what is going on and I could see the links being removed with the timeout.

> There isn't an elegant or straightforward way to do that. PipeWire is the one managing each filter state and there isn't a pipeline object with flushing mechanisms like in GStreamer. What we could try is some kind of hack. Creating a node like the one from the test signal that plays silence and linking it to our virtual sink. If that is going to have some unexpected side effect is hard to say without trying it.

Yeah, I didn't think this was going to be clean.
And your idea of implementation was effectively what I was asking for.
If you find this to be too janky/hacky, I understand.

**Comment by wwmm** [comment]
>Like when a node suspends with ["session.suspend-timeout-seconds"] = 5 ?

I am not sure. Depending on how complex the processing graph is PipeWire may or may not be able to automatically stop sending buffers to filters in pipelines whose players/recorders are inactive. We unlink the filters exactly as a workaround for the fact PipeWire is not always able to avoid wasting CPU in these situations. It is not clear to me if PipeWire is reading the timeout you are talking about when deciding if it has to stop calling filters callbacks or not.

>Yeah, I didn't think this was going to be clean.
>And your idea of implementation was effectively what I was asking for.
>If you find this to be too janky/hacky, I understand.

Compared to what is possible to do in GStreamer it is definitely not good. But it is not the first time that the lack of flushing has caused bad side effects. Eventually we will have to force some kind of flushing. I will try to find some time to investigate how hard it will be to put a hack solution in place.

**Comment by zDEFz** [comment]
This also fixed my issues!

like you said, `aplay -r 48000 -f S24_3LE /dev/zero`

**Comment by n3thshan** [comment]
is there any downside of using `aplay -r 48000 -f S24_3LE /dev/zero`? For example, higher battery usage on laptops? Is this the only viable workaround available? I was thinking of running this command on boot using Ignition.

**Comment by zDEFz** [comment]
> is there any downside of using `aplay -r 48000 -f S24_3LE /dev/zero`? For example, higher battery usage on laptops? Is this the only viable workaround available? I was thinking of running this command on boot using Ignition.

The other option is to not use easyeffects and use native parametric equalizer and stuff pipewire provides by itself.
Easyeffects would need to write the files for it if you want a GUI.
As for battery, it should be negligible.

**Comment by DarwinAwardWinner** [comment]
@n3thshan Yes, the downside is power/battery usage. Playing continuous silence on my laptop keeps EE running at around 5-7% CPU constantly, roughly doubling power usage since the CPU is never idle enough to sleep.

**Comment by DarwinAwardWinner** [comment]
So, I got fed up with the battery drain of constantly playing silence, and I wrote a program to play an additional second of silence after every other stream has stopped, and then stop, allowing the audio pipeline to suspend and save power when sound isn't actually playing Unfortuately and surprisingly, it turns out that this doesn't fix this issue. The only reason that continuously playing silence fixes it is because it causes the audio output to *never* suspend, which sidesteps the issue. By watching `pw-top` while testing this, I'm pretty sure that it's actually latency changes that are triggering the issue, not steam suspension and resumption.

Currently, the best solution I've found is somewhat unintuitive: set a very *short* inactivity timeout. I guess unloading and reloading the EE pipeline does somehow reset whatever stale state the Convolver plugin was holding onto. I set the inactivity timeout to 1s, so as long as I don't pause, seek, and then unpause all within a single second, I no longer head the crackling on unpause. I don't know if something else changed within EE recently that enables this workaround, but it actually works quite well for me now.

**Comment by wwmm** [comment]
>I guess unloading and reloading the EE pipeline does somehow reset whatever stale state the Convolver plugin was holding onto.

Yes. The filter instance is destroyed when easyeffects unlinks its pipeline. Any audio data stored inside the plugin is removed in this case. This can be good or bad. That is why we allow the user to customize the destruction timeout.

**Comment by DarwinAwardWinner** [comment]
Is there any downside to having a really short inactivity timeout? I haven't seen this suggested as a solution before, but it seems to pretty much completely solve my issue.

**Comment by wwmm** [comment]
> Is there any downside to having a really short inactivity timeout? I haven't seen this suggested as a solution before, but it seems to pretty much completely solve my issue.

As far as I remember only one. And you may or may not notice it. Imagine you paused a player for a few moments and resume playback. IF the plugins are destroyed the audio data inside them is lost. This inevitably causes a discontinuity in the audio signal. Some people notice this and don't like it. Other people like myself don't hear any artifact resulting from this. It depends on hardware and on the ear sensitivity. If you don't hear anything weird when you resume playback there is nothing for you to worry about.

**Comment by DarwinAwardWinner** [comment]
So basically, when you unpause after the plugins are suspended, you don't get exactly the same audio stream that you would get if the plugins remained loaded the whole time? If that's the case, I think the "constantly playing silence" workaround should also have the same issue. In any case, I'm not sure I'd even call it an issue. This seems like the correct behavior to me.

**Comment by wwmm** [comment]
>I think the "constantly playing silence" workaround should also have the same issue

This workaround does not allow easyeffects to destroy its pipeline. The older audio data will be eventually processed as silent audio buffers are received at the pipeline input.

**Comment by zDEFz** [comment]
> > I think the "constantly playing silence" workaround should also have the same issue
> 
> This workaround does not allow easyeffects to destroy its pipeline. The older audio data will be eventually processed as silent audio buffers are received at the pipeline input.

What I still do not get after so many weeks after me hinting to redo it.... why are there no attempts to use pipewires own configuration files instead?

**Comment by wwmm** [comment]
>why are there no attempts to use pipewires own configuration files instead?

For what exactly? Our filters are not based on pipewire's built-in filter chain. So if you have in mind settings for that it won't work. They do nothing for us.

**Comment by zDEFz** [comment]
> > why are there no attempts to use pipewires own configuration files instead?
> 
> For what exactly? Our filters are not based on pipewire's built-in filter chain. So if you have in mind settings for that it won't work. They do nothing for us.

The parametric filter for instance can completely get replaced by it, the link is provided above.
Using it completely avoids the noise issue.
The more these are used, the less issues are present...

**Comment by wwmm** [comment]
> The parametric filter for instance can completely get replaced by it, the link is provided above.
> Using it completely avoids the noise issue.
> The more these are used, the less issues are present...

The issue with PipeWire's built-in filter chain is that it was designed to be used inside pipewire. It is super inconvenient to use that outside of it. For example consider getting the level meters values. It is a huge annoyance. Controlling the parameters on the fly? Another huge annoyance.

It is the same drama that happens when a process has to communicate with another. The whole task requires a lot of code just because of the communication. And PipeWire's API for that is particularly obscure to understand. The few times I had to deal with it I cursed all the generations of whoever invented that thing 😄 . I am sorry but life is too precious to be wasted that way.
